db/sync.go

```diff
@@ -51,3 +54,78 @@ func (db *DB) BeginSync(batchPushURL string, diffServerURL string, dataLayerAuth
 
 	return syncHeadRef.TargetHash(), syncInfo, nil
 }
+
+func (db *DB) MaybeEndSync(syncHead hash.Hash) (bool, []Mutation, error) {
+	v := db.Noms().ReadValue(syncHead)
+	if v == nil {
+		return false, []Mutation{}, fmt.Errorf("could not load sync head %s", syncHead)
+	}
+	var syncHeadCommit Commit
+	if err := marshal.Unmarshal(v, &syncHeadCommit); err != nil {
+		return false, []Mutation{}, err
+	}
+
+	db.lock()
+	defer db.mu.Unlock()
+
+	// Stop if someone landed a sync since this sync started.
+	syncSnapshot, err := baseSnapshot(db.noms, syncHeadCommit)
+	if err != nil {
+		return false, []Mutation{}, err
+	}
+	syncSnapshotBasis, err := syncSnapshot.Basis(db.noms)
+	if err != nil {
+		return false, []Mutation{}, err
+	}
+	headSnapshot, err := baseSnapshot(db.noms, db.head)
+	if err != nil {
+		return false, []Mutation{}, err
+	}
+	if !syncSnapshotBasis.Original.Equals(headSnapshot.Original) {
+		return false, []Mutation{}, fmt.Errorf("sync aborted: found a newer snapshot %s on master", headSnapshot.Original.Hash())
+	}
+
+	// Determine if there are any pending mutations that we need to replay.
+	pendingCommits, err := pendingCommits(db.noms, db.head)
```

Comment:
[postfix] It seems like you could change `pendingCommits` to accept a `higherThan` parameter and pass `syncHeadCommit.MutationID()` in for that value, eliminating one line here plus the `filterIDsLessThanOrEqualTo` function.

Reply:
pendingCommits is an easy thing to understand and compose with, the caller and slice it and dice it how they want. i prefer this way myself though do not feel strongly.